src/gewis/service/gewisdb-service.ts

```diff
@@ -0,0 +1,128 @@
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+import GewisUser from '../entity/gewis-user';
+import { BasicApi, Configuration, Health, MembersApi } from 'gewisdb-ts-client';
+import log4js, { Logger } from 'log4js';
+import { webResponseToUpdate } from '../helpers/gewis-helper';
+import UserService from '../../service/user-service';
+import { UserResponse } from '../../controller/response/user-response';
+
+const GEWISDB_API_URL = process.env.GEWISDB_API_URL;
+const GEWISDB_API_KEY = process.env.GEWISDB_API_KEY;
+
+// Configuration for the API access
+const configuration = new Configuration({ basePath: GEWISDB_API_URL, accessToken: () => GEWISDB_API_KEY });
+const api = new MembersApi(configuration);
+const pinger = new BasicApi(configuration);
+
+// Logger setup
+const logger: Logger = log4js.getLogger('GewisDBService');
+logger.level = process.env.LOG_LEVEL;
+
+export default class GewisDBService {
+
+  public static api = api;
+
+  public static pinger = pinger;
+
+  /**
+   * Synchronizes ALL users with GEWIS DB user data.
+   * This method only returns users that were actually updated during the synchronization process.
+   * @returns {Promise<UserResponse[]>} A promise that resolves with an array of UserResponses for users that were updated. Returns null if the API is unhealthy.
+   */
+  public static async syncAll(): Promise<UserResponse[]> {
+    const gewisUsers = await GewisUser.find({ where: { user: { deleted: false } }, relations: ['user'] });
+    return this.sync(gewisUsers);
+  }
+
+  /**
+   * Synchronizes users with GEWIS DB user data.
+   * This method only returns users that were actually updated during the synchronization process.
+   * @param {GewisUser[]} gewisUsers - Array of users to sync.
+   * @returns {Promise<UserResponse[]>} A promise that resolves with an array of UserResponses for users that were updated. Returns null if the API is unhealthy.
+   */
+  public static async sync(gewisUsers: GewisUser[]): Promise<UserResponse[]> {
+    const ping: Health = await GewisDBService.pinger.healthGet().then(member => member)
+      .catch((error) => {
+        logger.warn('Failed to ping GEWIS DB', error);
+        return null;
+      });
+
+    if (ping.sync_paused) {
+      logger.warn('GEWISDB API paused, aborting.');
+      return null;
+    }
+
+    if (!ping.healthy) {
+      logger.warn('GEWISDB API unhealthy, aborting.');
+      return null;
+    }
+
+    const updates: UserResponse[] = [];
+    const promises = gewisUsers.map(user => GewisDBService.updateUser(user).then((u) => {
+      if (u) updates.push(u);
+    }));
+
+    return Promise.all(promises).then(() => updates).catch(() => null);
```

Comment:
I am very confused by this function. If a single `updateUser` call fails, then this promise will return `null` for all users? I get very confused with combining `await` with `then` and `catch`, even though it works. I have the same issue with line 78, because there you only push the promise if it succeeds. But if it fails, the whole `sync` function crashes I think? I do not fully understand it, so in my opinion to keep the code readable, use try-catch with `await` in the try block, or return a promise with a `then` and `catch` (which you do not await).

Reply:
alternative proposal with promise.allSettled